### PR TITLE
MOTECH-1659: Add support for float, short and char in MDS

### PR DESCRIPTION
**Pull request by atish160384**
* Database is migrated to V4
* Added Type validations for the new types
* Added form validations for the new types

### CODE REVIEW

File: platform/mds/mds/src/main/java/org/motechproject/mds/dto/TypeDto.java [hunk 1]
```diff
@@ -106,6 +106,27 @@
     );
 
     /**
+     * Constant <code>FLOAT</code> is a representation of the MDS Decimal type.
+     */
+    public static final TypeDto FLOAT = new TypeDto(
+            "mds.field.float", "mds.field.description.float", "flo", Float.class.getName()
```
Comment (tstalka) [inline platform/mds/mds/src/main/java/org/motechproject/mds/dto/TypeDto.java]: float

File: platform/mds/mds-web/src/main/resources/webapp/partials/widgets/field-edit-Value-floathtml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    <p ng-show="field.validation.criteria[3].enabled && field.validation.criteria[3].value">{{msg('mds.info.cannotBeInSet')}}&#58; <strong>{{field.validation.criteria[3].value}}</strong></p>
+</div>
+<span class="form-hint" ng-show="{{field.name}}.$error.float && !{{field.name}}.$pristine">{{msg('mds.error.incorrectValueDecimal')}}</span>
+<span class="form-hint" ng-show="{{field.name}}.$error.min && !{{field.name}}.$pristine && !{{field.name}}.$error.float">{{msg('mds.error.min')}}</span>
+<span class="form-hint" ng-show="{{field.name}}.$error.max && !{{field.name}}.$pristine && !{{field.name}}.$error.float">{{msg('mds.error.max')}}</span>
+<span class="form-hint" ng-class="{hide: checkDecimalValue(field.value, field.settings)}">
+    {{getInvalidDecimalMessage(field.settings)}}
+</span>
+<span class="form-hint" ng-show="( {{field.name}}.$error.insetNum || {{field.name}}.$error.outsetNum ) && !{{field.name}}.$pristine && !{{field.name}}.$error.float && !{{field.name}}.$error.min && !{{field.name}}.$error.max">
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/partials/widgets/field-edit-Value-floathtml]: remove space

File: platform/mds/mds-web/src/main/resources/webapp/partials/widgets/field-basic-defaultValue-short.html [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+       ng-model="field.basic.defaultValue" default-field-name-valid="{{field.basic.name}}"
+       min-validity="{{getValidationCriteria(field, 0)}}" max-validity="{{getValidationCriteria(field, 1)}}" ng-readonly="field.readOnly"
+       ng-class="{'has-error': {{field.basic.name}}.$invalid || checkMax(field.basic.defaultValue, field) || checkMin(field.basic.defaultValue, field) || checkInSet(field.basic.defaultValue, field) || checkOutSet(field.basic.defaultValue, field)}">
+
+<span class="form-hint" ng-show="{{field.basic.name}}.$error.integer">{{msg('mds.error.incorrectValueShort')}}</span>
+<span class="form-hint" ng-show="checkMin(field.basic.defaultValue, field) && !{{field.basic.name}}.$error.integer">{{msg('mds.error.min')}}</span>
+<span class="form-hint" ng-show="checkMax(field.basic.defaultValue, field) && !{{field.basic.name}}.$error.integer">{{msg('mds.error.max')}}</span>
+
+<span class="form-hint" ng-show="( checkInSet(field.basic.defaultValue, field) || checkOutSet(field.basic.defaultValue, field) ) && !checkMin(field.basic.defaultValue, field) && !checkMax(field.basic.defaultValue, field) && !{{field.basic.name}}.$error.integer">
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/partials/widgets/field-basic-defaultValue-short.html]: remove space

File: platform/mds/mds-web/src/main/resources/webapp/partials/widgets/field-edit-Value-floathtml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<input float-validity inset-validity="{{getValidationCriteria(field, 2)}}" outset-validity="{{getValidationCriteria(field, 3)}}"
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/partials/widgets/field-edit-Value-floathtml]: dot is missing before extension

File: platform/mds/mds/src/main/resources/db/migration/mysql/V3__MOTECH-1659.sql [hunk 5]
```diff
@@ -0,0 +1,50 @@
+--
```
Comment (tstalka) [inline platform/mds/mds/src/main/resources/db/migration/mysql/V3__MOTECH-1659.sql]: We need also migration for PostgreSQL

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 6]
```diff
@@ -2409,6 +2409,59 @@
...
+                        ctrl.$setValidity('float', true);
+                        originalValue = viewValue;
+                        viewValue = parseFloat(viewValue);
+                        if (isNaN(viewValue)) {
+                            viewValue = '';
+                        }
+                        if (TWOZERO_REGEXP.test(originalValue)) {
+                            setTimeout(function () {
+                                elm.val(viewValue);
+                            }, 1000);
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: Why you use timeout in validation?

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 7]
```diff
@@ -2409,6 +2409,59 @@
...
+                ctrl.$parsers.unshift(function(viewValue) {
+                if(viewValue === '' || CHAR_REGEXP.test(viewValue)) {
+                        ctrl.$setValidity('char', true);
+                        originalValue = viewValue;
+                        return viewValue;
+                     }
+                else {
+                    ctrl.$setValidity('char', false);
+                    return viewValue;
+                    }
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: incorrect indentation

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 8]
```diff
@@ -2409,6 +2409,59 @@
...
+                });
+            }
+        };
+    });
+
+    directives.directive('charValidity', function() {
+        var CHAR_REGEXP = new RegExp('^[A-z0-9]$');
+        return {
+        require: 'ngModel',
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: incorrect indentation

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 9]
```diff
@@ -2409,6 +2409,59 @@
...
+                        // it is invalid, return undefined (no model update)
+                        ctrl.$setValidity('float', false);
+                        return viewValue;
+                    }
+                });
+            }
+        };
+    });
+
+    directives.directive('charValidity', function() {
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: charValidation

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 10]
```diff
@@ -2409,6 +2409,59 @@
         };
     });
 
+    directives.directive('floatValidity', function() {
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: floatValidation

**Comment by tstalka** [comment]
Add new types to the mds-test-bundle entities and some IT tests

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 11]
```diff
@@ -2409,6 +2409,59 @@
...
+                        ctrl.$setValidity('float', false);
+                        return viewValue;
+                    }
+                });
+            }
+        };
+    });
+
+    directives.directive('charValidity', function() {
+        var CHAR_REGEXP = new RegExp('^[A-z0-9]$');
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: char is also - + _ etc. So i think you can use metacharacters (http://www.w3schools.com/jsref/jsref_obj_regexp.asp)

File: platform/mds/mds/src/main/resources/db/migration/mysql/V3__MOTECH-1659.sql [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+INSERT INTO Type
+SELECT id + 1, 'mds.field.description.character','mds.field.character','char','java.lang.Character'
+FROM Type
+ORDER BY id DESC
+LIMIT 1;
+
+-- Adding Type Validations to float
+
+INSERT INTO TypeValidation VALUES (12,'mds.field.validation.minValue',22), (13,'mds.field.validation.maxValue',22);
```
Comment (tstalka) [inline platform/mds/mds/src/main/resources/db/migration/mysql/V3__MOTECH-1659.sql]: Are you sure that these ids will be valid, maybe use a syntax that is above

**Comment by tstalka** [comment]
I can't find ticket for UUID, so you should also add it. 
@pgesek do we need more new types?

File: platform/mds/mds/src/main/resources/db/migration/mysql/V3__MOTECH-1659.sql [hunk 13]
```diff
@@ -0,0 +1,50 @@
+--
+-- Add new types to MDS
+--
+-- Adding Float to Type table
+--
+INSERT INTO Type
+SELECT id + 1, 'mds.field.description.float','mds.field.float','flo','java.lang.Float'
```
Comment (tstalka) [inline platform/mds/mds/src/main/resources/db/migration/mysql/V3__MOTECH-1659.sql]: float

**Comment by pgesek** [comment]
@tstalka @atish160384 I will create a separate ticket for UUID. Let's not cram to much into this.

I don't know which types we need - personally I don't need any additional types and I haven't heard any requests for new types. I'm pretty sure Chris was interested in the float type primarily. The types introduced in this PR are enough for me, we can create separate tickets for other types if they are required. (although the title of the PR is misleading, since from what I can see char is the only additional type)

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js (hunk 6 above)
Comment (atish160384) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: If user types more than one zero it makes it one zero after one second. It is already used for 'decimalValidity'.

**Comment by atish160384** [comment]
@tstalka Please review this.

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js (hunk 6 above)
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: Please avoid copy paste and add separate method for these validators

File: platform/mds/mds/src/test/java/org/motechproject/mds/it/reposistory/AllTypesContextIT.java [hunk 14]
```diff
@@ -113,6 +113,18 @@ public void shouldReturnDefaultData() throws Exception {
                 "mds.field.relationship.oneToOne", "mds.field.description.relationship.oneToOne", OneToOneRelationship.class.getName(),
                 asList("mds.form.label.cascadePersist", "mds.form.label.cascadeUpdate", "mds.form.label.cascadeDelete"), null
         );
+        assertType(
+        		"mds.field.float", "mds.field.description.float", Float.class.getName(),
```
Comment (tstalka) [inline platform/mds/mds/src/test/java/org/motechproject/mds/it/reposistory/AllTypesContextIT.java]: incorrect indentation

File: platform/mds/mds/src/test/java/org/motechproject/mds/it/reposistory/AllTypesContextIT.java [hunk 15]
```diff
@@ -113,6 +113,18 @@ public void shouldReturnDefaultData() throws Exception {
...
+                asList("mds.form.label.precision", "mds.form.label.scale"), asList("mds.field.validation.minValue", "mds.field.validation.maxValue", "mds.field.validation.mustBeInSet", "mds.field.validation.cannotBeInSet")
+        );
+        assertType(
+        		"mds.field.short", "mds.field.description.short", Short.class.getName(),
```
Comment (tstalka) [inline platform/mds/mds/src/test/java/org/motechproject/mds/it/reposistory/AllTypesContextIT.java]: incorrect indentation

File: platform/mds/mds/src/test/java/org/motechproject/mds/it/reposistory/AllTypesContextIT.java [hunk 16]
```diff
@@ -113,6 +113,18 @@ public void shouldReturnDefaultData() throws Exception {
...
+                null, asList("mds.field.validation.minValue", "mds.field.validation.maxValue", "mds.field.validation.mustBeInSet", "mds.field.validation.cannotBeInSet")
+        );
+        assertType(
+        		"mds.field.character", "mds.field.description.character", Character.class.getName(),
```
Comment (tstalka) [inline platform/mds/mds/src/test/java/org/motechproject/mds/it/reposistory/AllTypesContextIT.java]: incorrect indentation

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 17]
```diff
@@ -2409,6 +2409,59 @@
...
+            ctrl.$parsers.unshift(function(viewValue) {
+            if(viewValue === '' || CHAR_REGEXP.test(viewValue)) {
+                    ctrl.$setValidity('char', true);
+                    originalValue = viewValue;
+                    return viewValue;
+                 }
+            else {
+                ctrl.$setValidity('char', false);
+                return viewValue;
+                }
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: incorrect indentation

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 18]
```diff
@@ -2409,6 +2409,59 @@
...
+        return {
+        require: 'ngModel',
+        link: function(scope, element, attrs, ctrl) {
+            var elm = angular.element(element), originalValue;
+            ctrl.$parsers.unshift(function(viewValue) {
+            if(viewValue === '' || CHAR_REGEXP.test(viewValue)) {
+                    ctrl.$setValidity('char', true);
+                    originalValue = viewValue;
+                    return viewValue;
+                 }
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: incorrect indentation

File: ui/package.json [hunk 19]
```diff
@@ -22,7 +22,8 @@
     "nunjucks": "~2.2.0",
     "path": "~0.12.7",
     "require-dir": "~0.3.0",
-    "yargs": "~3.31.0"
+    "yargs": "~3.31.0",
+    "gulp-header": "~1.7.1"
```
Comment (tstalka) [inline ui/package.json]: For what we need this?

**Comment by tstalka** [comment]
Add new types fields to some entities from the mds-test-bundle entities and some simple IT test with them

**Comment by tstalka** [comment]
I created a simple entity with new fields and i wanted to create instance but some input was unloaded, check attached image.

newtypes

**Comment by atish160384** [comment]
Review this please

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js [hunk 20]
```diff
@@ -2381,6 +2381,41 @@
...
+        return {
+            require: 'ngModel',
+            link: function(scope, element, attrs, ctrl) {
+                var elm = angular.element(element), originalValue;
+                ctrl.$parsers.unshift(function(viewValue) {
+                    if (viewValue === '' || INTEGER_REGEXP.test(viewValue)) {
+                        // it is valid
+                        ctrl.$setValidity('short', true);
+                        originalValue = viewValue;
+                        viewValue = parseFloat(viewValue);
```
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: parseFloat in short type?

File: platform/mds/mds-web/src/main/resources/webapp/js/directives.js (hunk 6 above)
Comment (tstalka) [inline platform/mds/mds-web/src/main/resources/webapp/js/directives.js]: Please avoid copy paste

**Comment by tstalka** [comment]
test this psql

File: platform/mds/mds-test-bundle/src/main/java/org/motechproject/mds/test/domain/inheritancestrategies/Dog.java [hunk 21]
```diff
@@ -12,6 +12,9 @@
 
     @Field
     private int hiddenBones;
+    
+    @Field
+    private float age;
```
Comment (tstalka) [inline platform/mds/mds-test-bundle/src/main/java/org/motechproject/mds/test/domain/inheritancestrategies/Dog.java]: float age ? maybe call this field "weight"

**Comment by tstalka** [comment]
Looks good to me

**Comment by tstalka** [comment]
test this psql

**Pull request merged**